Libplanet/Blocks/BlockCommitExtensions.cs

```diff
@@ -0,0 +1,76 @@
+            this BlockCommit commit,
+            IEnumerable<PublicKey> validators) =>
+            validators.Count() * 2 / 3 < commit.Votes.Count(vote =>
+                vote.Flag == VoteFlag.Commit && vote.BlockHash.Equals(commit.BlockHash));
+
+        /// <summary>
+        /// Checks whether <see cref="BlockCommit"/> has same validator public key with given
+        /// <paramref name="validators"/>.
+        /// </summary>
+        /// <param name="commit">A <see cref="BlockCommit"/> to check.</param>
+        /// <param name="validators">A <see cref="IEnumerable{T}"/> of validator public key.</param>
+        /// <returns>Returns <see langword="true"/> if <see cref="BlockCommit"/> has same validator
+        /// public keys with <paramref name="validators"/>, otherwise returns
+        /// <see langword="false"/>. </returns>
+        public static bool HasSameValidators(
+            this BlockCommit commit,
+            IEnumerable<PublicKey> validators)
+        {
+            List<PublicKey> iaVoteToHashSetList = new List<PublicKey>();
+            foreach (var key in commit.Votes.Select(x => x.Validator))
+            {
+                iaVoteToHashSetList.Add(key);
+            }
+
+            ImmutableHashSet<PublicKey> voteValidators = iaVoteToHashSetList.ToImmutableHashSet();
+            var publicKeys = validators as PublicKey[] ?? validators.ToArray();
+            ImmutableHashSet<PublicKey> previousValidators = publicKeys.ToImmutableHashSet();
+
+            return iaVoteToHashSetList.Count == publicKeys.Count() &&
+                   voteValidators.SetEquals(previousValidators);
+        }
+
+        /// <summary>
+        /// Checks whether <see cref="BlockCommit"/> has valid signatures for every votes.
+        /// </summary>
+        /// <param name="commit">A <see cref="BlockCommit"/> to check.</param>
+        /// <returns>Returns <see langword="true"/> if <see cref="BlockCommit.Votes"/> has valid
+        /// signatures, otherwise return <see langword="false"/>.</returns>
+        /// <remarks>A vote with <see cref="VoteFlag.Null"/> or <see cref="VoteFlag.Unknown"/> is
+        /// excluded from target, so these case returns <see langword="true"/>.</remarks>
+        public static bool HasValidVotes(this BlockCommit commit) =>
+            commit.Votes.All(
+                vote => vote.Flag == VoteFlag.Null || vote.Flag == VoteFlag.Unknown ||
+                        vote.Verify());
+
+        /// <summary>
+        /// Checks whether <see cref="BlockCommit"/> has same height for every votes.
+        /// </summary>
+        /// <param name="commit">A <see cref="BlockCommit"/> to check.</param>
+        /// <returns>Returns <see langword="true"/> if <see cref="BlockCommit.Votes"/> have same
+        /// height, otherwise returns <see langword="false"/>.</returns>
+        public static bool HasVotesSameHeight(this BlockCommit commit) =>
+            commit.Votes.All(vote => vote.Height == commit.Height);
+    }
+}
```

Comment:
I don't have any strong opinion, just asking for your opinion.
Each `HasTwoThirdsCommits()`, `HasSameValidators()`, `HasValidVotes()` will perform iteration, so if we want to validate, it will take 3 times of iteration.
Compared to single validation method with multiple exception cases, this implementation is much clear but may needs more computation.
If we don't have to perform individual validation, may be integrating them would be also an option I think, though not sure if it's a better design.
(Current design is obviously clearer for reading)

Reply:
Yes, I do agree on the point this would have a performance hit, but I thought that getting a review before any optimization would be better.